src/service/action-impl.js

```diff
@@ -399,6 +399,18 @@ export function parseActionMap(s, context) {
   const assertAction = assertActionForParser.bind(null, s, context);
   const assertToken = assertTokenForParser.bind(null, s, context);
 
+  // Consumes all subsequent token pairs of the form "<.><ID>" and invokes
+  // `callback` for each <ID> token. Used for variable dereferences.
+  function tokenizeIdDereferences(tokenizer, callback) {
+    for (let peek = tokenizer.peek();
+        peek.type == TokenType.SEPARATOR && peek.value == '.';
+        peek = tokenizer.peek()) {
+      tokenizer.next(); // Skip '.'.
+      const token = assertToken(tokenizer.next(false), [TokenType.ID]);
+      callback(token);
```

Comment:
`join` iterates through the array, no?

Why is returning an array here better? TBH I don't have a strong opinion so I'll change it if you insist. 😃